devito/types/tensor.py

```diff
@@ -79,211 +128,84 @@ def __subfunc_setup__(cls, *args, **kwargs):
-        # All cases should be covered but defaults to sympy if in another case
-        else:
-            return super(TensorFunction, self).__mul__(other)
+        if getattr(other, 'is_DiscreteFunction', False):
+            tmp_func = Dummy(other.name)
+            simplemul = super(TensorFunction, self).__mul__(tmp_func)
+            return simplemul.subs(tmp_func, other)
+        return super(TensorFunction, self).__mul__(other)
 
-    def __rmul__(self, other):
+    def __getattr__(self, name):
         """
-        Right multiplication of a TensorFunction T with u (u*T)
-        where u can be a TensorFunction, a VectorFunction or a Function/scalar.
-        Computes it via __mul__ with its transpose (T^T*u^T)^T.
+        Try calling a dynamically created FD shortcut.
+
+        Notes
+        -----
+        This method acts as a fallback for __getattribute__
         """
-        other = sympy.sympify(other)
         try:
-            if other.is_Function:
-                return self._eval_scalar_mul(other)
-            elif other.is_VectorValued or other.is_TensorValued:
-                return (self.T.__mul__(other.T)).T
-            else:
-                return super(TensorFunction, self).__rmul__(other)
-        except AttributeError:
-            return super(TensorFunction, self).__mul__(other)
+            return self.applyfunc(lambda x: x if x == 0 else getattr(x, name))
+        except:
+            raise AttributeError("%r object has no attribute %r" % (self.__class__, name))
 
     def _eval_at(self, func):
         """
         Evaluate tensor at func location
         """
-        def entry(i, j):
+        def entries(i, j, func):
             return getattr(self[i, j], '_eval_at', lambda x: self[i, j])(func[i, j])
-        comps = [[entry(i, j) for i in range(self.cols)] for j in range(self.rows)]
-        to = getattr(self, 'time_order', 0)
-        func = tens_func(self, self)
-        return func(name='%s%s' % (self.name, func.name),
-                    grid=self.grid, space_order=self.space_order,
-                    components=comps, time_order=to, symmetric=self.is_symmetric,
-                    diagonal=self.is_diagonal)
-
-    @classmethod
-    def __dtype_setup__(cls, **kwargs):
-        return Function.__dtype_setup__(**kwargs)
+        entry = lambda i, j: entries(i, j, func)
+        return self._new(self.rows, self.cols, entry)
 
     @classmethod
     def __indices_setup__(cls, **kwargs):
         return Function.__indices_setup__(grid=kwargs.get('grid'),
                                           dimensions=kwargs.get('dimensions'))
 
     @property
-    def is_diagonal(self):
-        return self._is_diagonal
-
-    @property
-    def is_symmetric(self):
-        return self._is_symmetric
-
-    @property
-    def indices(self):
-        return self._indices
-
-    @property
-    def staggered(self):
-        return self._staggered
+    def is_TimeDependent(self):
```

Comment:
Reminder for next PR: this property is probably useless as soon as you rename the class attribute `_is_TimeDependent` as `is_TimeDependent`